aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move

```diff
@@ -22,58 +22,61 @@ module aptos_framework::aggregator_v2 {
 
     /// Adds `value` to aggregator.
     /// If addition would exceed the max_value, `false` is returned, and aggregator value is left unchanged.
-    public native fun try_add<Element>(aggregator: &mut Aggregator<Element>, value: Element): bool;
+    public native fun try_add<IntElement>(aggregator: &mut Aggregator<IntElement>, value: IntElement): bool;
 
     // Adds `value` to aggregator, uncoditionally.
     // If addition would exceed the max_value, EAGGREGATOR_OVERFLOW exception will be thrown
-    public fun add<Element>(aggregator: &mut Aggregator<Element>, value: Element) {
+    public fun add<IntElement>(aggregator: &mut Aggregator<IntElement>, value: IntElement) {
         assert!(try_add(aggregator, value), error::out_of_range(EAGGREGATOR_OVERFLOW));
     }
 
     /// Subtracts `value` from aggregator.
     /// If subtraction would result in a negative value, `false` is returned, and aggregator value is left unchanged.
-    public native fun try_sub<Element>(aggregator: &mut Aggregator<Element>, value: Element): bool;
+    public native fun try_sub<IntElement>(aggregator: &mut Aggregator<IntElement>, value: IntElement): bool;
 
     // Adds `value` to aggregator, uncoditionally.
     // If subtraction would result in a negative value, EAGGREGATOR_UNDERFLOW exception will be thrown
-    public fun sub<Element>(aggregator: &mut Aggregator<Element>, value: Element) {
+    public fun sub<IntElement>(aggregator: &mut Aggregator<IntElement>, value: IntElement) {
         assert!(try_sub(aggregator, value), error::out_of_range(EAGGREGATOR_UNDERFLOW));
     }
 
     /// Returns a value stored in this aggregator.
-    public native fun read<Element>(aggregator: &Aggregator<Element>): Element;
+    public native fun read<IntElement>(aggregator: &Aggregator<IntElement>): IntElement;
 
-    public native fun snapshot<Element>(aggregator: &Aggregator<Element>): AggregatorSnapshot<Element>;
+    public native fun snapshot<IntElement>(aggregator: &Aggregator<IntElement>): AggregatorSnapshot<IntElement>;
 
     public native fun create_snapshot<Element: copy + drop>(value: Element): AggregatorSnapshot<Element>;
 
     public native fun copy_snapshot<Element: copy + drop>(snapshot: &AggregatorSnapshot<Element>): AggregatorSnapshot<Element>;
 
     public native fun read_snapshot<Element>(snapshot: &AggregatorSnapshot<Element>): Element;
 
-    public native fun string_concat<Element>(before: String, snapshot: &AggregatorSnapshot<Element>, after: String): AggregatorSnapshot<String>;
+    /// Concatenates `before`, `snapshot` and `after` into a single string.
+    /// snapshot passed needs to have integer type - currently supported types are u64 and u128.
+    /// raises EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE if called with another type.
+    public native fun string_concat<IntElement>(before: String, snapshot: &AggregatorSnapshot<IntElement>, after: String): AggregatorSnapshot<String>;
```

Comment:
Why does the input snapshot have to only an integer snapshot? The concatenation logic could work even in the input snapshot is a string snapshot right?

Reply:
yes, code could work - but that could generate arbitrary nested dependency pattern, and is probably not too useful. 
we limit to integers here, so that nesting level is at most two - concat(snapshot(aggregator))